File: Tokenizer.cpp
bool charCheck(char c, string s) {
  for(char k : s)
    if(c == k)
      return true;
  return false;
}
int charToInt(char c) { return c - '0'; }
bool isWhitespace(char c) { return c==' ' or c=='\n' or c=='\t'; }
bool isAlpha(char c) { return 'a'<=c and c<='z' or 'A'<=c and c<='Z'; }
bool isNumeric(char c) { return '0'<=c and c<='9'; }
bool isHexadecimal(char c) { return isNumeric(c) or 'a'<=c and c<='f' or 'A'<=c and c<='F'; }
bool isAlphanumeric(char c) { return isAlpha(c) or isNumeric(c); }
bool isIdentifier(char c) { return isAlphanumeric(c); }
bool isParenthesis(char c) { return charCheck(c, "()[]{}"); }

pair<char, Token::Type> blockingCharacters[] = {
  {'[', Token::Type::VectorStart},
  {']', Token::Type::VectorEnd},
  {'{', Token::Type::LambdaStart},
  {'}', Token::Type::LambdaEnd},
  {'(', Token::Type::ParenthesisOpen},
  {')', Token::Type::ParenthesisClose},
  {',', Token::Type::VectorSeparator},
  {';', Token::Type::Pipe},
};
bool isBlockingCharacter(char c) {
  for(auto p : blockingCharacters)
    if(c == p.first)
      return true;
  return false;
}
pair<string, Token::Type> nonblockingCharacters[] = {
  {".", Token::Type::SequenceSeparator},
  {"@", Token::Type::MethodMark},
  {"|", Token::Type::LambdaSeparator},
};
bool isBinaryOperator(char c) {
  return c != EOF and !isWhitespace(c) and !isIdentifier(c) and
    !isBlockingCharacter(c) and c != '\'' and c != '"';
}
int charToHex(char a) {
  if('0'<=a and a<='9') return a-'0';
  else if('a'<=a and a<='f') return a-'a'+10;
  else if('A'<=a and a<='F') return a-'A'+10;
}
char bytesToByte(char a, char b) {
  return charToHex(a) * 16 + charToHex(b);
}
char escapeCharacter(char c) {
  switch(c) {
  case 'n': return '\n';
  case '\\': return '\\';
  case '\'': return '\'';
  case '\"': return '\"';    
  case 't': return '\t';
  case 'r': return '\r';
  default: return c;
  }
}


class Tokenizer;
class TokenizerException : public SolException {
public:
  TokenizerException(Tokenizer *tokenizer, string message);
  TokenizerException(Token token, string message);
};
class TokenizerState {
public:
  StreamState *streamState;
  Token token;
  TokenizerState(StreamState *streamState, Token token) :
    streamState(streamState), token(token) {}
};
class Tokenizer {
public:
  CharacterStream *stream;
  Token token;
  Tokenizer(CharacterStream *stream) : stream(stream) { nextToken(); }
  void close() { stream->close(); }
  bool hasTokens() { return token.type != Token::Type::Eof; }
  TokenizerState *saveState() { return new TokenizerState(stream->saveState(), token); }
  void restoreState(TokenizerState *state) {
    stream->restoreState(state->streamState);
    token = state->token;
  }
  Token getToken() { return token; }
  void nextToken() {
    
    while(skipCharacters());
    
    token.location = stream->currentLocation();
    if(stream->isEOF()) {
      token.type = Token::Type::Eof;
      return;
    }
    
    if(getNumber()) return;
    if(getString()) return;
    if(getCharacter()) return;
    if(getSymbol()) return;
    if(getIdentifierKeyword()) return;
    if(getComment()) return;
    if(getLineComment()) return;
    
    for(auto p : blockingCharacters) 
      if(getSingleCharacterToken(p.first, p.second))
	return;

    if(getBinaryOperator()) {
      for(auto p : nonblockingCharacters)
	if(token.binaryOperator == p.first) {
	  token.type = p.second;
	  return;
	}
      return;
    }
    throw TokenizerException(this, "Unrecognized character.");
  }
  bool skipCharacters() {
    if(getWhitespace())     return true;
    if(getLineComment())    return true;
    if(getComment())        return true;
    if(getReverseComment()) return true;
    return false;
  }
  bool getWhitespace() {
    if(isWhitespace(stream->getCharacter())) {
      stream->nextCharacter();
      return true;
    } else
      return false;
  }
  bool checkConstantString(string s) {
    StreamState *state = stream->saveState();
    for(char c : s)
      if(stream->getCharacter() == c)
	stream->nextCharacter();
      else {
	stream->restoreState(state);
	return false;
      }
    return true;
  }
  bool getReverseComment() {
    return checkConstantString("-}");
  }
  bool getComment() {
    if(checkConstantString("{-")) {
      while(!stream->isEOF() and !checkConstantString("-}"))
	stream->nextCharacter();
      return true;
    } else return false;
  }
  bool getLineComment() {
    if(checkConstantString("--")) {
      while(!stream->isEOF() and stream->getCharacter() != '\n')
	stream->nextCharacter();
      stream->nextCharacter();
      return true;
    } else return false;
  }
  bool getNumber() {
    if(!isNumeric(stream->getCharacter()))
      return false;
    int integerPart = 0;
    while(isNumeric(stream->getCharacter())) {
      integerPart *= 10;
      integerPart += charToInt(stream->getCharacter());
      stream->nextCharacter();
    }
    StreamState *state = stream->saveState();
    if(stream->getCharacter() == '.') {
      stream->nextCharacter();
      if(!isNumeric(stream->getCharacter())) {
	stream->restoreState(state);
	token.type = Token::Type::Integer;
	token.integerValue = integerPart;
	return true;
      }
      token.type = Token::Type::Double;
      token.doubleValue = integerPart;
      double doublePart = 0.0;
      int exp = -1;
      while(isNumeric(stream->getCharacter())) {
	token.doubleValue += charToInt(stream->getCharacter()) * pow(10,exp);
	exp--;
	stream->nextCharacter();
      }
    } else {
      token.type = Token::Type::Integer;
      token.integerValue = integerPart;
    }
    return true;
  }
  bool getString() {
    if(stream->getCharacter() != '"')
      return false;
    token.type = Token::Type::String;
    token.stringValue = "";
    while(stream->getCharacter() == '"') {
      stream->nextCharacter();
      while(stream->getCharacter() != '"') {
	if(stream->getCharacter() == EOF)
	  throw TokenizerException(this, "Unexpected EOF in string.");
	else if(stream->getCharacter() == '\n') {
	  stream->nextCharacter();
	  skipCharacters();
	} else if(stream->getCharacter() == '\\') {
	  token.stringValue += parseEscapeSequence();
	  stream->nextCharacter();
	} else {	
	  token.stringValue += stream->getCharacter();
	  stream->nextCharacter();
	}
      }
      stream->nextCharacter();
      skipCharacters();
    }
    return true;      
  }
  char parseEscapeSequence() {
    stream->nextCharacter();
    char a = stream->getCharacter();
    if(a == EOF) throw TokenizerException(this, "Expected escape character value but got EOF.");
    if(a == '\n') throw TokenizerException(this, "Expected escape character value but got newline.");
    if(isHexadecimal(a)) {
      stream->nextCharacter();
      char b = stream->getCharacter(); 
      if(b == EOF) throw TokenizerException(this, "Expected hexadecimal character value but got EOF.");
      if(b == '\n') throw TokenizerException(this, "Expected hexadecimal character value but got newline.");
      if(!isHexadecimal(b))
	throw TokenizerException(this, "Expected valid hexadecimal character value.");
      return bytesToByte(a, b);
    } else
      return escapeCharacter(stream->getCharacter());
  }
  bool getCharacter() {
    if(stream->getCharacter() != '\'')
      return false;
    stream->nextCharacter();
    token.type = Token::Type::Character;
    if(stream->getCharacter() == '\n')
      throw TokenizerException(this, "Expected character but got newline.");
    else if(stream->getCharacter() == EOF)
      throw TokenizerException(this, "Expected character but got EOF.");
    else if(stream->getCharacter() == '\\')
      token.characterValue = parseEscapeSequence();
    else
      token.characterValue = stream->getCharacter();	 
    stream->nextCharacter();
    if(stream->getCharacter() != '\'')
      throw TokenizerException(this, "Expected closing character quote.");
    stream->nextCharacter();
    return true;      
  }
  bool getSymbol() {
    if(stream->getCharacter() != '#')
      return false;
    stream->nextCharacter();
    token.type = Token::Type::Symbol;
    token.symbol = "";
    if(isIdentifier(stream->getCharacter()))
      while(isIdentifier(stream->getCharacter()) or stream->getCharacter() == ':') {
	token.symbol += stream->getCharacter();
	stream->nextCharacter();
      }
    else if(isBinaryOperator(stream->getCharacter()))
      while(isBinaryOperator(stream->getCharacter())) {
	token.symbol += stream->getCharacter();
	stream->nextCharacter();
      }
    else
      throw TokenizerException(this, "Expected valid identifier character in symbol.");
    return true;
  }
  bool getSingleCharacterToken(char character, Token::Type type) {
    if(stream->getCharacter() != character)
      return false;
    token.type = type;
    stream->nextCharacter();
    return true;
  }
  bool getIdentifierKeyword() { // Either Identifier or Keyword
    if(!isIdentifier(stream->getCharacter()))
      return false;
    string symbol;
    while(isIdentifier(stream->getCharacter())) {
      symbol += stream->getCharacter();
      stream->nextCharacter();
    }
    if(stream->getCharacter() == ':') {
      stream->nextCharacter();
      token.type = Token::Type::Keyword;
      token.keyword = symbol;
    }
    else {
      token.type = Token::Type::Identifier;
      token.identifier = symbol;
    }
    return true;
  }
  bool getBinaryOperator() {
    if(!isBinaryOperator(stream->getCharacter()))
      return false;
    token.type = Token::Type::BinaryOperator;
    token.binaryOperator = "";
    while(isBinaryOperator(stream->getCharacter())) {
      token.binaryOperator += stream->getCharacter();
      stream->nextCharacter();
    }
    return true;
  }
  Tokenizer *debugPrint() {
    TokenizerState *state = saveState();
    while(hasTokens()) {
      cout<<getToken().toString()<<endl;
      nextToken();
    }
    restoreState(state);
    return this;
  }
};

TokenizerException::TokenizerException(Tokenizer *tokenizer, string message) :
  SolException("Tokenizer", tokenizer->getToken().location, message) {}
TokenizerException::TokenizerException(Token token, string message) :
  SolException("Tokenizer", token.location, message) {}

File: Tokenizer.java
import java.io.*;
import java.util.*;

class TokenizerException extends SolException {
    TokenizerException(Tokenizer tokenizer, String message) throws TokenizerException {
	super("Tokenizer", tokenizer.getToken().getLocation(), message);
    }
    TokenizerException(Tokenizer tokenizer, Token token, String message) {
	super("Tokenizer", token.getLocation(), message);
    }
    TokenizerException(Tokenizer tokenizer, Location location, String message) {
	super("Tokenizer", location, message);
    }
}
public class Tokenizer {
    CharacterStream stream;
    Token token;

    class State {
	public CharacterStream.State state;
	public Token token;
	State(CharacterStream.State state, Token token) {
	    this.state = state;
	    this.token = token;
	}
    }
    Tokenizer(CharacterStream stream) throws TokenizerException {
	this.stream = stream;
    }
    public State saveState() {
	return new State(stream.saveState(), token);
    }
    public void restoreState(State state) {
	stream.restoreState(state.state);
	token = state.token;
    }
    public static boolean isIdentifier(char c) {
	return Character.isLetterOrDigit(c);
    }
    public static boolean isMessageBlockingCharacter(char c) {
	return "[]{}(),;".indexOf(c) != -1;
    }
    public static boolean isBinaryOperator(char c) {
	return c != 0 && !Character.isWhitespace(c) && !isIdentifier(c) &&
	    !isMessageBlockingCharacter(c) && c != '\'' && c != '"';
    }
    public void close() { stream.close(); }
    public boolean hasTokens() throws TokenizerException { return !(getToken() instanceof EOFToken); }
    public Token getToken() throws TokenizerException {
	if(token == null)
	    nextToken();
	return token;
    }
    void nextToken() throws TokenizerException {
	skipCharacters();
	Location loc = new Location(stream.getLocation());
	if(stream.hasCharacters())
	    nextTokenNonlocalized();
	else
	    token = new EOFToken();	     
	token.setLocation(loc);
    }
    void nextTokenNonlocalized() throws TokenizerException {
	Location loc = new Location(stream.getLocation());
		
	if((token = getNumber())            != null) return;
	if((token = getString())            != null) return;
	if((token = getCharacter())         != null) return;
	if((token = getMessage())            != null) return;
	if((token = getIdentifierOrKeyword()) != null) return;
	
	if(getSingleCharacterToken('[')) { token = new VectorStartToken();      return; }
	if(getSingleCharacterToken(']')) { token = new VectorEndToken();        return; }
	if(getSingleCharacterToken('{')) { token = new LambdaStartToken();      return; }
	if(getSingleCharacterToken('}')) { token = new LambdaEndToken();        return; }
	if(getSingleCharacterToken('(')) { token = new ParenthesisOpenToken();  return; }
	if(getSingleCharacterToken(')')) { token = new ParenthesisCloseToken(); return; }
	if(getSingleCharacterToken(',')) { token = new VectorSeparatorToken();  return; }
	if(getSingleCharacterToken(';')) { token = new PipeToken();             return; }
	
	if((token = getBinaryOperator()) != null) {
	    BinaryOperatorToken binaryOperator = (BinaryOperatorToken)token;
	    if(binaryOperator.getBinaryOperator().equals("."))
		token = new SequenceSeparatorToken();
	    if(binaryOperator.getBinaryOperator().equals("@"))
		token = new MethodMarkToken();
	    if(binaryOperator.getBinaryOperator().equals("|"))
		token = new LambdaSeparatorToken();
	    return;
	}
	throw new TokenizerException(this, loc, "Unrecognized character.");	
    }
    public boolean getSingleCharacterToken(char character) {
	if(stream.getCharacter() != character)
	    return false;
	stream.nextCharacter();
	return true;
    }
    public void skipCharacters() {
	while(skipWhitespace()
	      || skipLineComment()
	      || skipComment()
	      || skipReverseComment());
    }
    public boolean skipWhitespace() {
	if(Character.isWhitespace(stream.getCharacter())) {
	    stream.nextCharacter();
	    return true;
	} else
	    return false;
    }
    public boolean skipConstantString(String s) {
	CharacterStream.State state = stream.saveState();
	for(char c : s.toCharArray())
	    if(stream.getCharacter() == c)
		stream.nextCharacter();
	    else {
		stream.restoreState(state);
		return false;
	    }
	return true;
    }
    public boolean skipReverseComment() {
	return skipConstantString("-}");
    }
    public boolean skipComment() {
	if(skipConstantString("{-")) {
	    while(stream.hasCharacters() && !skipConstantString("-}"))
		stream.nextCharacter();
	    return true;
	} else return false;
    }
    public boolean skipLineComment() {
	if(skipConstantString("--")) {
	    while(stream.hasCharacters() && stream.getCharacter() != '\n')
		stream.nextCharacter();
	    stream.nextCharacter();
	    return true;
	} else return false;
    }
    ///////////////////////////////////////////////////////////////
    public Token getNumber() {
	if(!Character.isDigit(stream.getCharacter()))
	    return null;
	int integerValue = 0;
	while(Character.isDigit(stream.getCharacter())) {
	    integerValue *= 10;
	    integerValue += Character.getNumericValue(stream.getCharacter());
	    stream.nextCharacter();
	}
	CharacterStream.State state = stream.saveState();
	if(stream.getCharacter() == '.') {
	    stream.nextCharacter();
	    if(!Character.isDigit(stream.getCharacter())) {
		stream.restoreState(state);
		return new IntegerToken(integerValue);
	    }
	    double doubleValue = 0.0;
	    int exp = -1;
	    while(Character.isDigit(stream.getCharacter())) {
		doubleValue +=
		    Character.getNumericValue(stream.getCharacter()) * Math.pow(10,exp);
		exp--;
		stream.nextCharacter();
	    }
	    return new DoubleToken(integerValue + doubleValue);
	} else
	    return new IntegerToken(integerValue);
    }
    public Token getString() throws TokenizerException {
	if(stream.getCharacter() != '"')
	    return null;
	String stringValue = "";
	while(stream.getCharacter() == '"') {
	    stream.nextCharacter();
	    while(stream.getCharacter() != '"') {
		if(!stream.hasCharacters())
		    throw new TokenizerException(this, "Unexpected EOF in string.");
		if(stream.getCharacter() == '\n') {
		    stream.nextCharacter();
		    skipCharacters();
		} else if(stream.getCharacter() == '\\') {
		    stringValue += parseEscapeSequence();
		    stream.nextCharacter();
		} else {	
		    stringValue += stream.getCharacter();
		    stream.nextCharacter();
		}
	    }
	    stream.nextCharacter();
	    skipCharacters();
	}
	return new StringToken(stringValue);
    }
    public char parseEscapeSequence() throws TokenizerException {
	stream.nextCharacter();
	char a = stream.getCharacter();
	if(!stream.hasCharacters()) throw new TokenizerException(this, "Expected escape character value but got EOF.");
	if(a == '\n')      throw new TokenizerException(this, "Expected escape character value but got newline.");
	if(Character.digit(a, 16) != -1) {
	    stream.nextCharacter();

	    char b = stream.getCharacter(); 
	    if(!stream.hasCharacters()) throw new TokenizerException(this, "Expected hexadecimal character value but got EOF.");
	    if(b == '\n')      throw new TokenizerException(this, "Expected hexadecimal character value but got newline.");
	    if(Character.digit(a, 16) != -1)
		return (char)(Character.digit(a, 16) * 16 + Character.digit(b, 16));
	    else
		throw new TokenizerException(this, "Expected valid hexadecimal character value.");
	} else
	    return escapeCharacter(stream.getCharacter());
    }
    public static char escapeCharacter(char c) {
	switch(c) {
	case 'n':  return '\n';
	case '\\': return '\\';
	case '\'': return '\'';
	case '\"': return '\"';    
	case 't':  return '\t';
	case 'r':  return '\r';
	default:   return c;
	}
    }
    public Token getCharacter() throws TokenizerException {
	if(stream.getCharacter() != '\'')
	    return null;
	stream.nextCharacter();
	char character;
	if(!stream.hasCharacters())
	    throw new TokenizerException(this, "Expected character but got EOF.");
	else if(stream.getCharacter() == '\n')
	    throw new TokenizerException(this, "Expected character but got newline.");
	else if(stream.getCharacter() == '\\')
	    character = parseEscapeSequence();
	else
	    character = stream.getCharacter();	 
	stream.nextCharacter();
	if(stream.getCharacter() != '\'')
	    throw new TokenizerException(this, "Expected closing character quote.");
	stream.nextCharacter();
	return new CharacterToken(character);
    }
    public Token getMessage() throws TokenizerException {
	if(stream.getCharacter() != '#')
	    return null;
	CharacterStream.State state = stream.saveState();
	stream.nextCharacter();
	if(isIdentifier(stream.getCharacter())) {
	    ArrayList<String> parts = new ArrayList<>();
	    String part = "";
	    while(isIdentifier(stream.getCharacter()) || stream.getCharacter() == ':') {
		if(stream.getCharacter() == ':') {
		    if(part.equals(""))
			throw new TokenizerException(this, "Invalid empty keyword while parsing Message.");
		    parts.add(part);
		    part = "";
		} else
		    part += stream.getCharacter();
		stream.nextCharacter();
	    }
	    if(!part.equals(""))
		if(parts.size() == 0)
		    return new MessageToken(new UnaryMessage(part));
		else
		    throw new TokenizerException(this, "Expected keyword end.");
	    else
		return new MessageToken(new KeywordMessage(parts.toArray(new String[parts.size()])));
	} else if(isBinaryOperator(stream.getCharacter())) {
	    String message = "";
	    while(isBinaryOperator(stream.getCharacter())) {
		message += stream.getCharacter();
		stream.nextCharacter();
	    }
	    return new MessageToken(new BinaryMessage(message));
	} else {
	    stream.restoreState(state);
	    return null;
	}
    }
    public Token getIdentifierOrKeyword() {
	if(!isIdentifier(stream.getCharacter()))
	    return null;
	String symbol = "";
	while(isIdentifier(stream.getCharacter())) {
	    symbol += stream.getCharacter();
	    stream.nextCharacter();
	}
	if(stream.getCharacter() == ':') {
	    stream.nextCharacter();
	    return new KeywordToken(symbol);
	} else
	    return new IdentifierToken(symbol);
    }
    public Token getBinaryOperator() {
	if(!isBinaryOperator(stream.getCharacter()))
	    return null;
	String binaryOperator = "";
	while(isBinaryOperator(stream.getCharacter())) {
	    binaryOperator += stream.getCharacter();
	    stream.nextCharacter();
	}
	return new BinaryOperatorToken(binaryOperator);
    }
}
